File: apps/android/core/common/src/main/kotlin/com/sigmundgranaas/turbo/expressive/core/common/Outcome.kt
package com.sigmundgranaas.turbo.expressive.core.common

/**
 * A lightweight success/failure result for repository + use-case boundaries,
 * so the UI can render Loading/Content/Error without exceptions leaking up.
 */
sealed interface Outcome<out T> {
    data class Success<T>(val value: T) : Outcome<T>
    data class Failure(val error: Throwable) : Outcome<Nothing>

    fun getOrNull(): T? = (this as? Success)?.value

    fun <R> map(transform: (T) -> R): Outcome<R> = when (this) {
        is Success -> Success(transform(value))
        is Failure -> this
    }

    companion object {
        inline fun <T> catching(block: () -> T): Outcome<T> =
            try {
                Success(block())
            } catch (t: Throwable) {
                Failure(t)
            }
    }
}

File: apps/ios/Sources/CoreCommon/Outcome.swift
import Foundation

/// A lightweight success/failure result for repository + use-case boundaries,
/// so the UI can render Loading/Content/Error without exceptions leaking up.
///
/// Mirrors `core.common.Outcome` in the Android app. We keep a bespoke type
/// (rather than Swift's `Result`) so both clients read the same way.
public enum Outcome<T> {
    case success(T)
    case failure(Error)

    public func getOrNil() -> T? {
        switch self {
        case .success(let value): value
        case .failure: nil
        }
    }

    public func map<R>(_ transform: (T) -> R) -> Outcome<R> {
        switch self {
        case .success(let value): .success(transform(value))
        case .failure(let error): .failure(error)
        }
    }

    /// Run `block`, capturing any thrown error as a `.failure`.
    public static func catching(_ block: () throws -> T) -> Outcome<T> {
        do {
            return .success(try block())
        } catch {
            return .failure(error)
        }
    }

    /// Async variant of ``catching(_:)`` for `await`-ing repositories.
    public static func catching(_ block: () async throws -> T) async -> Outcome<T> {
        do {
            return .success(try await block())
        } catch {
            return .failure(error)
        }
    }
}

extension Outcome: Sendable where T: Sendable {}
